packages/types/src/tx/schema/tx.ts

```diff
@@ -17,9 +18,9 @@ export class TxMsgValue {
   chainId!: string;
 
   @field({ type: option("string") })
-  publicKey!: string | undefined;
+  publicKey?: string;
 
-  constructor(data: TxMsgValue) {
+  constructor(data: TxProps) {
```

Comment:
As mentioned in other comment, I think these are actually the same type.

```typescript
type IsSame =
  TxMsgValue extends TxProps
    ? TxProps extends TxMsgValue
      ? true
      : false
    : false;

const isSame: IsSame = true;
```

But it might make sense to keep both to distinguish between plain object and class, I'm not sure.

Reply:
Ah yes sorry I forgot to answer that one :D So with typescript it is fine, but then what happens is that borsh.ts reads `data.constructor` which has to be `TxMsgValue`. Otherwise it "can't find" the schema to parse data.
So we could use only TxMsgValue, but we need to remember about setting prototype again after deserialization.

Reply:
Ahh right I see. It seems `InstanceType<typeof TxMsgValue>` will just resolve to `TxMsgValue` though, and doesn't check the constructor. The `InstanceType` way still allowed me to pass in a plain object that is not an instance of `TxMsgValue`.

``` typescript
class Child {}

class Parent {
  child: InstanceType<typeof Child>;

  constructor(child: Child) {
    if (!(child instanceof Child)) {
      throw new Error("uh oh");
    }
    console.log("all good");
    this.child = child;
  }
}

new Parent(new Child()); // works fine
new Parent({}); // throws error at runtime
```

There doesn't look like a way to check the constructor at compile-time in TypeScript, so maybe there is nothing we can do here.

Reply:
Yeah I guess what we could do is remove props and change fields in schemas to private? :D This way we can force object construction using constructor xd I think what we have is good for now, Justin mentioned that he wants to remove nested "tx" in schemas anyway :P